### ISSUE TITLE
Hidden annotation with no replies is appearing in client

### ISSUE THREAD

**Issue by robertknight** [hypothesis/product-backlog#583]
From @gmaciocci at eLife:

> Noticing some off-spec bevaviour on comment moderation: “Hidden” comments with no replies were meant to disappear, but they stick around in redacted mode instead.  https://hyp.is/_zEedDmqEeidO4MgnHKWag/elifesciences.org/articles/33399. The expected behaviour is that unless a hidden annotation has replies which are not themselves hidden, it goes away from public view

This is how it currently appears:
![image](https://user-images.githubusercontent.com/502186/39762840-ad8b3d12-52a9-11e8-885a-0ec8e14dbb14.png)

Instead, the annotation should completely disappear from view.

**Comment by seanh** [comment]
@sheetaluk @ajpeddakotla did we decide to remove this card from the sprint?

**Comment by sheetaluk** [comment]
We can filter out the hidden annotations in one of 3 places: 
(1) The client
(2) In `h` possibly in the `views/api.py search`
(3) In ES.
The team has decided that the best place to do it would be in ES. Possibly add a new field `should_hide` while indexing an annotation into ES and then write filters for that field. The reason for doing it in ES is that its faster and also aggregation counts (eg. numbers of annotations with a particular tags), estimates of total number of results etc. which are also computed by ES remain correct.

**Comment by klemay** [comment]
Seeing similar behavior at http://www.dartmouth.edu/~milton/reading_room/pl/book_1/text.shtml:

<img width="1512" alt="screen shot 2018-07-30 at 12 48 51 pm" src="https://user-images.githubusercontent.com/12087705/43411096-304906c4-93f7-11e8-912a-d4035fc83480.png">

These annotations did have replies, but I flagged and hid the replies first. 

This comes out of a ticket I've been working on with @jeremydean related to a Dartmouth edX course.

**Comment by jeremydean** [comment]
This is currently a bad look with a major potential partner. Dartmouth has expressed interest in the LMS pilot and is currently piloting H within the edX platform. @dwhly 

After talking with instructor and designer for the course, we decided to remove public annotations that 1) were accidentally created on the public channel by previous students in previous courses and 2) the instructor felt distracted from the reading/use of the text in the current course (see screen shot above of an annotation that grabs too much text). See additional context at ZenDesk:

https://hypothesis.zendesk.com/agent/tickets/2754

@klemay attempted to do this through the moderation feature, but this bug is preventing the actual "deletion" of these annotations/solution for the partner.

@ajpeddakotla can we consider adding this to a sprint soon? looks like @sheetaluk already has a possible fix.

**Comment by klemay** [comment]
^ given this is creating an issue for a potential partner, I'm planning to bring this up with my meeting with Arti tomorrow morning (she is offline for the day by now)

**Comment by sheetaluk** [comment]
Additional notes:
This issue seems to happen when all children of an annotation AND the annotation itself, have the `nipsa` field set to `true`.
According to the `NipsaFilter` in `h/search`, it looks like the initial implementation was to account for showing a nipsad/hidden/moderated annotation in its redacted state if it had replies.
What we now need to do is check if the annotation has replies and if all of those replies has `nipsa: true` then filter that annotation from the search results.

The best thing to do is at the time of indexing, introduce a check to set a new field `should_hide` which will be set to `true` if that annotation has `nipsa: true` and all the annotations in the `thread_ids` array also have `nipsa: true` set.
Once this is done, we can create a filter in `query.py` similar to the `DeletedFilter` which will filter out annotations which has `should_hide: true` set.

This can seem a little confusing because it might seem like we are doing the showing/hiding of annotaitons in multiple places. But I can't think of a better way of doing it. Also, we should do this is ES because doing it in `h` or `client` will throw off the aggregations etc.

**Comment by sheetaluk** [comment]
This is wrapped inside the `replace_nipsa_with_hidden_filter` feature flag, so I think if its working as expected for staff, we could turn it on for everyone.
Once we confirm that its working as expected, we should create a followup issue to remove the feature flag and do the necessary cleanup.
